File: src/SimpleLevelEditor.Formats.Types/Level/Level3dData.fs
namespace SimpleLevelEditor.Formats.Types.Level

type Level3dData =
    { mutable EntityConfigPath: string option
      mutable ModelPaths: string list
      mutable WorldObjects: WorldObject list
      mutable Entities: Entity list }

    static member CreateDefault() =
        { EntityConfigPath = None
          ModelPaths = []
          WorldObjects = []
          Entities = [] }

    member this.DeepCopy() =
        { this with
            ModelPaths = this.ModelPaths |> List.map id
            WorldObjects = this.WorldObjects |> List.map (_.DeepCopy())
            Entities = this.Entities |> List.map (_.DeepCopy()) }

    member this.AddModel(model: string) =
        if not (this.ModelPaths |> List.exists (fun v -> v = model)) then
            this.ModelPaths <- this.ModelPaths @ [model]

    member this.AddWorldObject(worldObject: WorldObject) =
        this.WorldObjects <- this.WorldObjects @ [worldObject]

    member this.AddEntity(entity: Entity) =
        this.Entities <- this.Entities @ [entity]

    member this.RemoveModel(model: string) =
        this.ModelPaths <- this.ModelPaths |> List.filter (fun v -> v <> model)

    member this.RemoveWorldObject(worldObject: WorldObject) =
        this.WorldObjects <- this.WorldObjects |> List.filter (fun v -> v <> worldObject)

    member this.RemoveEntity(entity: Entity) =
        this.Entities <- this.Entities |> List.filter (fun v -> v <> entity)

File: src/SimpleLevelEditor.Formats.Level/Level3dData.cs
using System.Diagnostics.CodeAnalysis;

namespace SimpleLevelEditor.Formats.Level;

public sealed record Level3dData
{
	public required string? EntityConfigPath { get; set; }

	public required List<string> ModelPaths { get; set; }

	public required List<WorldObject> WorldObjects { get; set; }

	public required List<Entity> Entities { get; set; }

	[SetsRequiredMembers]
	public Level3dData(string? entityConfigPath, List<string> modelPaths, List<WorldObject> worldObjects, List<Entity> entities)
	{
		EntityConfigPath = entityConfigPath;
		ModelPaths = modelPaths;
		WorldObjects = worldObjects;
		Entities = entities;
	}

	public static Level3dData CreateDefault()
	{
		return new Level3dData(null, [], [], []);
	}

	public Level3dData DeepCopy()
	{
		return new Level3dData(EntityConfigPath, [..ModelPaths], [..WorldObjects.Select(wo => wo.DeepCopy())], [..Entities.Select(e => e.DeepCopy())]);
	}

	public void AddModel(string modelPath)
	{
		if (ModelPaths.Contains(modelPath))
			return;

		ModelPaths.Add(modelPath);
	}

	public void AddWorldObject(WorldObject worldObject)
	{
		WorldObjects.Add(worldObject);
	}

	public void AddEntity(Entity entity)
	{
		Entities.Add(entity);
	}

	public void RemoveModel(string modelPath)
	{
		ModelPaths.Remove(modelPath);
	}

	public void RemoveWorldObject(WorldObject worldObject)
	{
		WorldObjects.Remove(worldObject);
	}

	public void RemoveEntity(Entity entity)
	{
		Entities.Remove(entity);
	}
}
